plugin/src/main/java/com/google/errorprone/xplat/checker/J2objcMethodName.java

```diff
@@ -0,0 +1,326 @@
+
+import com.google.common.base.Splitter;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.ErrorProneFlags;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.bugpatterns.BugChecker.ClassTreeMatcher;
+import com.google.errorprone.bugpatterns.BugChecker.CompilationUnitTreeMatcher;
+import com.google.errorprone.bugpatterns.BugChecker.MethodTreeMatcher;
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.matchers.Matchers;
+import com.google.errorprone.matchers.MethodVisibility.Visibility;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.AnnotationTree;
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.CompilationUnitTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.tree.VariableTree;
+import com.sun.tools.javac.code.Symbol;
+import com.sun.tools.javac.code.Symbol.ClassSymbol;
+import com.sun.tools.javac.code.Type;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import javax.lang.model.type.TypeKind;
+
+/**
+ * Checks for Java methods that would become unreasonably long when translated by J2ObjC into a
+ * function. If a method has 6+ parameters and its name would be 80+ characters, a warning is given.
+ * By default, any method which would have a name with 300+ characters is warned against. This can
+ * be changed with the command line argument {@code -XepOpt:J2ObjCMethodName:MethodNameLength=X}
+ * where X is the minimum number of characters to warn on. For example, {@code
+ * -XepOpt:J2ObjCMethodName:MethodNameLength=100} will warn against all methods that would result in
+ * 100 or more characters after being translated to an Objective-C function.
+ */
+@BugPattern(
+    name = "J2ObjCMethodName",
+    summary = "Warns against long J2ObjC translated methods.",
+    explanation =
+        "J2ObjC translates Java methods in a specific format that can lead to unreasonably long"
+            + " Objective-C functions when the Java method has many parameters and/or uses long"
+            + " types. This checker calls out problematic methods and offers a fix.",
+    severity = WARNING)
+public class J2objcMethodName extends BugChecker implements MethodTreeMatcher,
+    ClassTreeMatcher, CompilationUnitTreeMatcher {
+
+  private final Map<String, String> foundObjcClassNames = new HashMap<>();
+
+  private final Map<String, Integer> usedMethodNames = new HashMap<>();
+
+  private String packageAnnotation;
+
+  private int methodNameLength = 300;
+
+  private static final Matcher<MethodTree> MATCHER =
+      Matchers.anyOf(
+          Matchers.isStatic(),
+          Matchers.methodHasVisibility(Visibility.PRIVATE)
```

Comment:
Shouldn't this be not private?

Reply:
I believe that static and private methods are translated into C functions, public methods are translated into Objective-C methods. I was under the impression that C functions are an issue and Objective-C methods are not, but I could be wrong.

Reply:
No mostly the C functions are unused and only the Object-C functions are used.   (Lukhnos can correct me if I'm wrong but if I am then my understanding of the process needs to be improved)

Reply:
It's true that native code won't have access (more precisely visibility) to the translated private methods—those are turned to C functions visible to the compilation unit only. The issue is really only with public static methods.

Reply:
Wouldn't this affect all public methods both static or ones that are members?

Reply:
> Wouldn't this affect all public methods both static or ones that are members?

Not really—non-static member methods only have Objective-C method translations and `@ObjectiveCName` does not affect those either. A public static Java method is always translated to a (Objective-)C function and by default also as an Objective-C class method (but can be stripped by invoking `j2objc --no-wrapper-methods`, as some projects actually do).